### PR TITLE
Refactor MakeWay

### PR DESCRIPTION
**Pull request by fmatthew5876**
~~Depends on #1628~~

This currently passes all test cases in #1675 ~~Except:~~

- [x] 1.2 - Collision parts pass, but the order of messages goes wrong later on. This is a deeper non-makeway related issue

Fix: #1675
Fix: #1612
Fix: #1606
Fixes a game hang reported in: #985

Retested: #884 
Retested: #1082

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This is now ready for review. 

I believe I have covered every possible collision case for player and vehicles here.

**Comment by carstene1ns** [comment]
Better diff: https://github.com/fmatthew5876/Player/compare/event_collide...fmatthew5876:event_makeway

**Comment by Ghabry** [comment]
Even the better diff is very hard to follow, too much code moved around everywhere.

**Comment by fmatthew5876** [comment]
For this one I basically ripped out every collision related function in `game_map.cpp`.

Therefore for reviewing, I would suggest maybe just reading the final version of `game_map.cpp` and only using diffs for the other files touched.

**Comment by fmatthew5876** [comment]
I added a commit that not only fixed #937, but also test case 1.2 from #1675 mentioned in the original post!

**Comment by Ghabry** [comment]
When you ignore the diff and open old and new file instead and do full text search you notice that it's actually not that bad, mostly is really just code moved around (Sometimes with minor additions because of new behaviour matthew figured out), e.g.:

- Map::AirshipLandOk & Vehicle::CanLand & Map::IsLandable -> Map::CanLandAirship
- Vehicle::Update (partly moved to) -> Game_Vehicle::AnimateAscentDescent
- Map::IsPassableVehicle is now integrated in Map::IsPassableTile
- New func Map::IsPassableLowerTile uses code ripped out of Map::IsPassableTile

The Jump code was mostly rewritten, but this is a local thing so no big deal to review this.

So the ones with real changes are WouldCollide, MakeWayCollideEvent and MakeWay, looking deeper into them.

File: src/game_map.cpp [hunk 1]
```diff
@@ -572,202 +572,226 @@ static CollisionResult TestCollisionDuringMove(
...
-
-	// A character can move to a position with an impassable tile by
-	// standing on top of an event below it. These flags track whether
-	// we stepped off an event and therefore don't need to check the
-	// passability of the tile layer below.
-	bool stepped_off_event = false;
-	bool stepped_onto_event = false;
-
-	for (Game_Event& other : GetEvents()) {
-		CollisionResult result = TestCollisionDuringMove(x, y, new_x, new_y, d, self, other);
```
Comment (Ghabry) [inline src/game_map.cpp]: Seems that TestCollisionDuringMove and the CollisionResult enum are now dead code and can be removed?

File: src/game_event.cpp [hunk 2]
```diff
@@ -521,11 +521,9 @@ void Game_Event::Update() {
...
 	}
 
-	if (IsProcessed()) {
-		return;
-	}
-	SetProcessed(true);
-
+	// RPG_RT runs the parallel interpreter everytime Update is called.
+	// That means if the event updates multiple times due to makeway,
+	// the interpreter will run multiple times per frame.
```
Comment (Ghabry) [inline src/game_event.cpp]: > Run parallel events everytime update is called 

Could you also add a comment that this also emulates a bug because the Wait count also decrements

File: src/game_map.cpp [hunk 3]
```diff
@@ -572,202 +572,226 @@ static CollisionResult TestCollisionDuringMove(
...
-		Output::Warning("IsPassableVehicle: Invalid terrain at (%d, %d)", x, y);
-	} else if (vehicle_type == Game_Vehicle::Boat) {
-		if (!terrain->boat_pass)
-			return false;
-	} else if (vehicle_type == Game_Vehicle::Ship) {
-		if (!terrain->ship_pass)
-			return false;
-	} else if (vehicle_type == Game_Vehicle::Airship) {
-		return terrain->airship_pass;
+		Output::Warning("MakeWay: Invalid terrain at (%d, %d)", x, y);
```
Comment (Ghabry) [inline src/game_map.cpp]: Warning msg: MakeWay -> CanLandAirship

File: src/game_map.cpp [hunk 4]
```diff
@@ -794,6 +818,80 @@ bool Game_Map::IsPassableTile(int bit, int tile_index) {
...
+bool Game_Map::IsPassableTile(const Game_Character* self, int bit, int x, int y) {
+	if (!IsValid(x, y)) return false;
+
+	auto vehicle_type = (self != nullptr)
+		? self->GetVehicleType() : Game_Vehicle::None;
+
+	if (vehicle_type != Game_Vehicle::None) {
+		const auto* terrain = ReaderUtil::GetElement(Data::terrains, GetTerrainTag(x, y));
+		if (!terrain) {
+			Output::Warning("MakeWay: Invalid terrain at (%d, %d)", x, y);
```
Comment (Ghabry) [inline src/game_map.cpp]: Warning msg: MakeWay -> IsPassableTile

File: src/game_event.cpp (hunk 2 above)
Comment (fmatthew5876) [reply src/game_event.cpp]: The wait count decrementing is a result of running the interpreter again. Do you still want a special comment for that?

Comment (Ghabry) [reply src/game_event.cpp]: because the wait is in frames and not in interpreter invocations I don't find this intuitive so yes, explicitly mentioning this is better.

Note, old comment was:
> This results in waits to finish quicker when an event collides with this event and
> emulates a RPG Maker bug)

File: src/game_map.cpp (hunk 1 above)
Comment (fmatthew5876) [reply src/game_map.cpp]: Removed

File: src/game_map.cpp (hunk 3 above)
Comment (fmatthew5876) [reply src/game_map.cpp]: Fixed

File: src/game_map.cpp (hunk 4 above)
Comment (fmatthew5876) [reply src/game_map.cpp]: Fixed

File: src/game_event.cpp (hunk 2 above)
Comment (fmatthew5876) [reply src/game_event.cpp]: Added

**Comment by Ghabry** [comment]
Maybe three things that could be retested are:

- [ ] #1002
- [ ] #1079
- [ ] #1349

**Comment by fmatthew5876** [comment]
I retested all those games. All passed.
